### PR TITLE
Revert SyncTransport simplification and fix properly

### LINKED ISSUE #34115: [Messenger][BC-break] Sync transport can no longer be configured by env variable
**Symfony version(s) affected**: 4.4.x-dev
**Description**  
Since https://github.com/symfony/symfony/pull/34069 the `sync://` transport is handled differently. 
The case when the transport name is defined using an environment variable is broken.

Works:
```
framework:
   messenger:
     transports:
       default: "sync://"
```

Broken case:
```
parameters:
  env(MESSENGER_TRANSPORT): sync://

framework:
   messenger:
     transports:
       default: "%env(MESSENGER_TRANSPORT)%"
```

Yields `No transport supports the given Messenger DSN "sync://"` when dispatching a message.

**How to reproduce**  
See https://github.com/Devristo/sf4.4-messenger-bug for a reproducer.

**Possible Solution**  

 * Document this BC-break
 * Fix the BC-break: If someone could give me some pointers I am willing to implement a fix.

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #34115 (and also related to #34066)
| License       | MIT
| Doc PR        | Not needed

In #34069, I made `SyncTransport` simpler by removing that transport class and making the whole things a config trick. I felt GREAT about that solution... until i realized two big problems:

1) It kills using env vars for `sync://` because we read the config values at build time - #34115 - that could probably be fixed by adding a factory, but then there is also the next problem

2) If someone routed a message to `[async, sync]` (weird, but allowed), my #34069 config solution basically maps this internally to `[async]`, which actually causes the message to *not* be handled immediately. Basically, my solution only worked if you route a message ONLY to one sync transport, but fails if you route to multiple transports.

So... this fixes things in a less-cool, but sensible way:

A) The first commit reverts #34069 exactly
B) The second commit solves the issue that we need to know if a message is being handled in a "worker" context or not, so middleware can decide if they should reset things before/after handling things. Previously we were using `ReceivedStamp` to know this. But because `SyncTransport` also "receives" the message and adds this stamp, it's not enough. To fix this, I added a new `ConsumedByWorkerStamp` that clearly means: "This message is being handled by a worker" (and so, you might want to "reset" some things before/after handling).

Thanks!

### CODE REVIEW

File: src/Symfony/Component/Messenger/Worker.php [hunk 1]
```diff
@@ -132,7 +133,7 @@ private function handleMessage(Envelope $envelope, ReceiverInterface $receiver,
         ];
 
         try {
-            $envelope = $this->bus->dispatch($envelope->with(new ReceivedStamp($transportName)));
+            $envelope = $this->bus->dispatch($envelope->with(new ReceivedStamp($transportName), new ConsumedByWorkerStamp()));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Worker.php]: Is there anything the worker might want to pass to the chain? E.g. the `$output` of the command? I'm just wondering :)

**Review by nicolas-grekas — APPROVED** [review]
Makes sense, although it's strange to have two ways to send things "sync".
Maybe the other way to be sync is the one that should be removed?
Anyway, barely related :)

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/Messenger/Worker.php (hunk 1 above)
Comment (weaverryan) [reply src/Symfony/Component/Messenger/Worker.php]: Even though this stamp won't be "sent" on redelivery, it feels odd to pass what's really a "service" into a Stamp. Better might be (if someone needs this) to add it to one of the worker events we dispatch.

Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Worker.php]: I see no issues on my side doing so , the day we need to.

**Comment by weaverryan** [comment]
Only failure is because `symfony/doctrine-bridge:dev-master` tests are looking for the new `ConsumedByWorkerStamp` inside `symfony/messenger:dev-master`, but it won't be there until this is merged up to `master`. But if this is fixable by a `composer.json` constraint, let me know.

**Review by Tobion — APPROVED** [review]

**Comment by Tobion** [comment]
Thank you @weaverryan.

**Pull request merged**